pkg/policy/types/types.go

```diff
@@ -58,4 +63,53 @@ func (k Key) PortProtoIsEqual(c Key) bool {
 	return k.DestPort == c.DestPort && k.Nexthdr == c.Nexthdr
 }
 
+// PrefixLength returns the prefix lenth of the key
+// for indexing it.
+func (k Key) PrefixLength() uint {
+	p := MapStatePrefixLen
+	if k.DestPort == 0 {
+		p -= 16
+		// We can only mask Nexthdr
+		// if DestPort is also masked.
+		if k.Nexthdr == 0 {
+			p -= 8
+		}
+	}
+	return p
+}
+
+// CommonPrefix implements the CommonPrefix method for the
+// bitlpm.Key interface. Identity is not indexed and is instead,
+// saved as a simple map per TrafficDirection-Protocol-Port index
+// key.
+func (k Key) CommonPrefix(b Key) uint {
+	v := bits.LeadingZeros8(k.TrafficDirection ^ b.TrafficDirection)
+	if v != 8 {
+		return uint(v)
+	}
+	v += bits.LeadingZeros8(k.Nexthdr ^ b.Nexthdr)
+	if v != 16 {
+		return uint(v)
+	}
+	return uint(v + bits.LeadingZeros16(k.DestPort^b.DestPort))
+}
+
+// BitValueAt implements the BitValueAt method for the
+// bitlpm.Key interface.
+func (k Key) BitValueAt(i uint) uint8 {
+	if i < 8 {
+		return min(k.TrafficDirection&(1<<(7-i)), 1)
+	}
+	if i < 16 {
+		return min(k.Nexthdr&(1<<(7-(i-8))), 1)
+	}
+	return uint8(min(k.DestPort&(1<<(15-(i-16))), 1))
+}
```

Comment:
I guess `Key.Identity` is zeroed on insert so that the `Value` method will return a Key with zero Identity? Otherwise the zeroing of the `Identity` does not seem significant?

Reply:
> I guess Key.Identity is zeroed on insert so that the Value method will return a Key with zero Identity? Otherwise the zeroing of the Identity does not seem significant?

It's done for unit testing purposes. Outside of tests, it does not provide any functionality.